In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_excel('D:/DocumentsD/4to_semestre/Analisis de Ciencia de Datos/Reto-Ternium/ODEs_1año.xlsx')

In [3]:
print(df.columns)

Index(['ID_ODE', 'T_USUARIO_ALTA', 'T_USUARIO_ULT_MOD', 'ESTADO_ODE',
       'FECHA_ALTA_ODE', 'CONSIGNATARIO', 'SOLICITANTE', 'C_ID_REMITO',
       'FECHA_DESP', 'C_ITEM', 'ESPESOR_DESP', 'ANCHO_DESP', 'LARGO_DESP',
       'PESO_NETO_DESP', 'PESO_BRUTO_DESP', 'C_POSICION_ORDEN', 'C_PEDIDO',
       'MATERIAL_PADRE', 'PESO_MATERIAL_ENTRADA', 'ANCHO_PADRE', 'LARGO_PADRE',
       'ESPESOR_PADRE', 'ID_MAT', 'MATERIAL', 'PESO_MATERIAL_SALIDA',
       'ANCHO_HIJO', 'LARGO_HIJO', 'ESPESOR_HIJO', 'DESCRIPCION_EMBALAJE',
       'C_OFA_ID', 'D_UBICACION', 'NUMERO_PARTE_HIJO', 'PIEZAS_POR_PAQ_HIJO',
       'PESO_MAX_DESPACHO_HIJO', 'ID_RED_HIJO', 'ARMA_TARIMA', 'FECHA_ARMADO',
       'USUARIO_ARMA', 'DESARMA_TARIMA', 'FECHA_DESARME', 'USUARIO_DESARMA'],
      dtype='str')


['ID_ODE', 'CONSIGNATARIO','PESO_NETO_DESP',
                    'MATERIAL_PADRE',
                    'ANCHO_HIJO', 'LARGO_HIJO', 'ESPESOR_HIJO',
                    'DESCRIPCION_EMBALAJE',
                    'NUMERO_PARTE_HIJO','PIEZAS_POR_PAQ_HIJO',
                    'PESO_MAX_DESPACHO_HIJO','ARMA_TARIMA',
                    'DESARMA_TARIMA']

                    #D_UBICACION

Quitamos todos los datos que no tengan valor en arma tarima ya que no se les armó una tarima por lo tanto no nos interesan 

In [4]:
df = df.dropna(subset=['ARMA_TARIMA'])

In [5]:
filas_restantes = len(df)
print(filas_restantes)

27743


En el dataframe "dF" dejamos unicamente las variables que nos sirven para identificar los patrones entre las variables que se quieren probar


In [6]:
variables_utiles = ['ID_ODE', 'CONSIGNATARIO','SOLICITANTE','C_ITEM','PESO_BRUTO_DESP','C_PEDIDO',
                    'MATERIAL_PADRE',
                    'PESO_MATERIAL_SALIDA','DESCRIPCION_EMBALAJE',
                    'NUMERO_PARTE_HIJO','PIEZAS_POR_PAQ_HIJO',
                    'PESO_MAX_DESPACHO_HIJO','ARMA_TARIMA',
                    'DESARMA_TARIMA','D_UBICACION','ANCHO_HIJO', 'LARGO_HIJO', 'ESPESOR_HIJO',]



df = df[variables_utiles]

In [7]:
df = df.sort_values('ARMA_TARIMA', ascending=False)
df

,ID_ODE,CONSIGNATARIO,SOLICITANTE,C_ITEM,PESO_BRUTO_DESP,C_PEDIDO,MATERIAL_PADRE,PESO_MATERIAL_SALIDA,DESCRIPCION_EMBALAJE,NUMERO_PARTE_HIJO,PIEZAS_POR_PAQ_HIJO,PESO_MAX_DESPACHO_HIJO,ARMA_TARIMA,DESARMA_TARIMA,D_UBICACION,ANCHO_HIJO,LARGO_HIJO,ESPESOR_HIJO
70583,23152673,N000100452,N000100452,1,3.234,301690063.0,4A620748CD300,3226,EC52 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180326AB,NOA,NOA,TAR480769-APT,NaN,Churubusco,327.80,369.863,3.390
70584,23152673,N000100452,N000100452,2,3.234,301690063.0,4A620748CD300,3226,EC52 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180326AB,NOA,NOA,TAR480769-APT,NaN,Churubusco,327.80,369.863,3.390
70579,23152673,N000100452,N000100452,1,3.316,301690063.0,4A620888CD300,3308,EC52 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180326AB,NOA,NOA,TAR480763-APT,NaN,Churubusco,327.87,380.306,3.380
70580,23152673,N000100452,N000100452,2,3.316,301690063.0,4A620888CD300,3308,EC52 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180326AB,NOA,NOA,TAR480763-APT,NaN,Churubusco,327.87,380.306,3.380
70626,23135759,N000110899,N000100521,3,1.378,301688784.0,4A470962CD300,1372,EC57 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180687AA,NOA,NOA,TAR480601-APT,TAR480601-APT,Churubusco,255.00,202.807,3.380
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
8041,22012869,N000124490,N000100779,4,1.556,300926467.0,4A171086CD300,1550,EC52 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,RM2022-200317-ERMA,NOA,13.00,TAR343176-APT,TAR343176-APT,Churubusco,198.11,488.631,2.040
34680,22528059,N000110899,N000100521,8,0.601,301637297.0,4A272719JP200,600,EC03 - CINTA/PAPEL/OJO VERTICAL/TERRESTRE,NaN,NOA,NOA,TAR343154-APT,NaN,Juventud,133.40,1263.820,0.455
34681,22528059,N000110899,N000100521,9,0.631,301637297.0,4A272719JP201,630,EC03 - CINTA/PAPEL/OJO VERTICAL/TERRESTRE,NaN,NOA,NOA,TAR343154-APT,NaN,Juventud,133.40,1327.120,0.455
56804,22923370,N000127299,H010520452,2,0.684,301672398.0,2A989808UG400,677,EC03 - CINTA/PAPEL/OJO VERTICAL/TERRESTRE,RMLG205100,NOA,NOA,TAR283167-EMBSL2_JUV,TAR283167-APT,Juventud,130.54,750.279,0.883


In [8]:
df.head()

,ID_ODE,CONSIGNATARIO,SOLICITANTE,C_ITEM,PESO_BRUTO_DESP,C_PEDIDO,MATERIAL_PADRE,PESO_MATERIAL_SALIDA,DESCRIPCION_EMBALAJE,NUMERO_PARTE_HIJO,PIEZAS_POR_PAQ_HIJO,PESO_MAX_DESPACHO_HIJO,ARMA_TARIMA,DESARMA_TARIMA,D_UBICACION,ANCHO_HIJO,LARGO_HIJO,ESPESOR_HIJO
70583,23152673,N000100452,N000100452,1,3.234,301690063.0,4A620748CD300,3226,EC52 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180326AB,NOA,NOA,TAR480769-APT,NaN,Churubusco,327.80,369.863,3.39
70584,23152673,N000100452,N000100452,2,3.234,301690063.0,4A620748CD300,3226,EC52 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180326AB,NOA,NOA,TAR480769-APT,NaN,Churubusco,327.80,369.863,3.39
70579,23152673,N000100452,N000100452,1,3.316,301690063.0,4A620888CD300,3308,EC52 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180326AB,NOA,NOA,TAR480763-APT,NaN,Churubusco,327.87,380.306,3.38
70580,23152673,N000100452,N000100452,2,3.316,301690063.0,4A620888CD300,3308,EC52 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180326AB,NOA,NOA,TAR480763-APT,NaN,Churubusco,327.87,380.306,3.38
70626,23135759,N000110899,N000100521,3,1.378,301688784.0,4A470962CD300,1372,EC57 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180687AA,NOA,NOA,TAR480601-APT,TAR480601-APT,Churubusco,255.00,202.807,3.38


Se quiere ver si una orden la satisfacen difetentes plantas o solo una la satisface

In [22]:
grupos = df.groupby('ID_ODE')['D_UBICACION'].nunique()

# Grupos donde ID_ODE tiene más de un valor distinto de C_PEDIDO
inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada ID_ODE siempre tiene el mismo valor de D_UBICACION en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} ID_ODE(s) tienen más de un valor distinto en D_UBICACION:")
    print(inconsistentes)

CONFIRMADO: Cada ID_ODE siempre tiene el mismo valor de D_UBICACION en todas sus filas.


Se quiere probar que todas las cintas que pertenecen a una misma orden pertenecen a un mismo C_PEDIDO

In [23]:
# Para cada ID_ODE, verificar si todos sus valores de C_PEDIDO son iguales
grupos = df.groupby('ID_ODE')['C_PEDIDO'].nunique()

# Grupos donde ID_ODE tiene más de un valor distinto de C_PEDIDO
inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada ID_ODE siempre tiene el mismo valor de C_PEDIDO en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} ID_ODE(s) tienen más de un valor distinto en C_PEDIDO:")
    print(inconsistentes)


CONFIRMADO: Cada ID_ODE siempre tiene el mismo valor de C_PEDIDO en todas sus filas.


Cada orden es entregada a donde mismo 

In [24]:
grupos = df.groupby('ID_ODE')['CONSIGNATARIO'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada ID_ODE siempre tiene el mismo valor de CONSIGNATARIO en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} ID_ODE(s) tienen más de un valor distinto en CONSIGNATARIO:")
    print(inconsistentes)

CONFIRMADO: Cada ID_ODE siempre tiene el mismo valor de CONSIGNATARIO en todas sus filas.


Se quiere probar si una orden la encarga mas de un cliente

In [25]:
grupos = df.groupby('ID_ODE')['SOLICITANTE'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada ID_ODE siempre tiene el mismo valor de SOLICITANTE en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} ID_ODE(s) tienen más de un valor distinto en SOLICITANTE:")
    print(inconsistentes)

CONFIRMADO: Cada ID_ODE siempre tiene el mismo valor de SOLICITANTE en todas sus filas.


Se quiere ver si un cliente ha realizado varias ordenes a ternium lo cual es cierto

In [26]:
grupos = df.groupby('SOLICITANTE')['ID_ODE'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada SOLICITANTE siempre tiene el mismo valor de ID_ODE en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} SOLICITANTE(s) tienen más de un valor distinto en ID_ODE:")
    print(inconsistentes)

NO SE CUMPLE: 101 SOLICITANTE(s) tienen más de un valor distinto en ID_ODE:
SOLICITANTE
3000003209     5
3000004302    10
3000004576    73
6800000026    27
6800000108    14
              ..
N000100867    23
N000190807    40
N000191206     9
N000191581     2
N000191748     5
Name: ID_ODE, Length: 101, dtype: int64


Queremos ver si en una misma orden se piden varios "productos" o tambien llamados "combitos", lo que es cierto, en una misma orden el cliente puede pedir varios productos

In [27]:
grupos = df.groupby('ID_ODE')['NUMERO_PARTE_HIJO'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada ID_ODE siempre tiene el mismo valor de NUMERO_PARTE_HIJO en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} ID_ODE(s) tienen más de un valor distinto en NUMERO_PARTE_HIJO:")
    print(inconsistentes)

NO SE CUMPLE: 55 ID_ODE(s) tienen más de un valor distinto en NUMERO_PARTE_HIJO:
ID_ODE
20813371     2
20875103     5
20878598     2
21964071     5
22038080     8
22038125     4
22038134     2
22133727     2
22208945     2
22213463     2
22302106     2
22302312     2
22302571     2
22326613     3
22326616     2
22333028    10
22469214     2
22469819     2
22521903     2
22527968     2
22528059     6
22528811     6
22529289     2
22529414     2
22540391     3
22620489     2
22620530     2
22709041     3
22712997     3
22715739     3
22726060     3
22748190     2
22769576     2
22783594     2
22887592     3
22923370     5
22923847     2
22924718     2
22931131     2
22931194     5
22931207     2
22931894     2
22937888     2
22937904     4
22937906     2
22937913     4
22937921     2
22937929    15
22952809     4
22969399     2
22969408     2
23025273     3
23113560    15
23135759     2
23147293     3
Name: NUMERO_PARTE_HIJO, dtype: int64


Queremos ver si los dichos "combitos" o "prodcutos" son unicos para el cliente, es decir si 2 clientes diferentes pueden pedir o han pedido el mismo producto, lo cual es cierto y esto prueba que "NUMERO_PARTE_HIJO" es un id del producto que vende ternium y este producto es comprado por diferentes clientes

In [28]:
grupos = df.groupby('NUMERO_PARTE_HIJO')['ID_ODE'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada NUMERO_PARTE_HIJO siempre tiene el mismo valor de ID_ODE en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} NUMERO_PARTE_HIJO(s) tienen más de un valor distinto en ID_ODE:")
    print(inconsistentes)

NO SE CUMPLE: 550 NUMERO_PARTE_HIJO(s) tienen más de un valor distinto en ID_ODE:
NUMERO_PARTE_HIJO
003-0816-00                       2
003-0900-99                       3
003-1143-00                       2
003-1511-00                       2
003-1721-04                       4
                                 ..
W11681023                         2
Z CINTA 3' C.20 G33 ANCH CONF    29
Z CINTA 3' C.22 G33 G90 ANCH     26
Z CINTA 3' C.24 G33 ANCH CONF     4
Z CINTA 4' C.26 G33 ANCH CONF     2
Name: ID_ODE, Length: 550, dtype: int64


Como se explica en el markdown de arriba esto prueba que un producto es comprado por diferentes clientes

In [29]:
grupos = df.groupby('NUMERO_PARTE_HIJO')['SOLICITANTE'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada NUMERO_PARTE_HIJO siempre tiene el mismo valor de SOLICITANTE en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} NUMERO_PARTE_HIJO(s) tienen más de un valor distinto en SOLICITANTE:")
    print(inconsistentes)

NO SE CUMPLE: 67 NUMERO_PARTE_HIJO(s) tienen más de un valor distinto en SOLICITANTE:
NUMERO_PARTE_HIJO
005-1312-00                      2
01-16230-48                      2
01-16891-39                      2
01-16891-41                      2
01-16996-03                      2
                                ..
S05H042719600                    2
W10496916                        2
Z CINTA 3' C.20 G33 ANCH CONF    5
Z CINTA 3' C.22 G33 G90 ANCH     4
Z CINTA 3' C.24 G33 ANCH CONF    2
Name: SOLICITANTE, Length: 67, dtype: int64


Con este codigo queremos ver si un producto es conformado por diferente material padre, lo que es cierto, cada prodcuto es conformado por diferentes tipo de material padre 

In [30]:
grupos = df.groupby('NUMERO_PARTE_HIJO')['MATERIAL_PADRE'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada NUMERO_PARTE_HIJO siempre tiene el mismo valor de MATERIAL_PADRE en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} NUMERO_PARTE_HIJO(s) tienen más de un valor distinto en MATERIAL_PADRE:")
    print(inconsistentes)

NO SE CUMPLE: 549 NUMERO_PARTE_HIJO(s) tienen más de un valor distinto en MATERIAL_PADRE:
NUMERO_PARTE_HIJO
003-0816-00                       5
003-0900-99                       3
003-1511-00                       2
003-1721-04                       2
004287-12186                      4
                                 ..
Z CINTA 3' C.20 G33 ANCH CONF    31
Z CINTA 3' C.22 G33 ANCH CONF     2
Z CINTA 3' C.22 G33 G90 ANCH     18
Z CINTA 3' C.24 G33 ANCH CONF     4
Z CINTA 4' C.26 G33 ANCH CONF     5
Name: MATERIAL_PADRE, Length: 549, dtype: int64


Se prueba que Un mismo tipo de material padre sirve para realizar distintos prodcutos 

In [31]:
grupos = df.groupby('MATERIAL_PADRE')['NUMERO_PARTE_HIJO'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada MATERIAL_PADRE siempre tiene el mismo valor de NUMERO_PARTE_HIJO en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} MATERIAL_PADRE(s) tienen más de un valor distinto en NUMERO_PARTE_HIJO:")
    print(inconsistentes)

NO SE CUMPLE: 433 MATERIAL_PADRE(s) tienen más de un valor distinto en NUMERO_PARTE_HIJO:
MATERIAL_PADRE
2B251041VG500    2
2B251041VG501    2
2B289925VG501    2
2B289925VG502    2
2B342619VG500    2
                ..
4A617277PD300    2
4A619788PD300    2
4A619831UG201    2
4A620853PD300    2
4A621664PD300    2
Name: NUMERO_PARTE_HIJO, Length: 433, dtype: int64


Se crea un nuevo dataframe ordenado por ordenes para ver cómo es que se realiza una orden y como se termina por desarmar las tarimas

In [32]:
df_desarme = df
variables_compaDesarme = ['ID_ODE','C_ITEM',
                    'MATERIAL_PADRE',
                    'DESCRIPCION_EMBALAJE',
                    'NUMERO_PARTE_HIJO','PIEZAS_POR_PAQ_HIJO',
                    'ARMA_TARIMA',
                    'DESARMA_TARIMA']



df_desarme = df_desarme[variables_compaDesarme]
df_desarme = df_desarme.sort_values('ID_ODE', ascending=False)
df_desarme

,ID_ODE,C_ITEM,MATERIAL_PADRE,DESCRIPCION_EMBALAJE,NUMERO_PARTE_HIJO,PIEZAS_POR_PAQ_HIJO,ARMA_TARIMA,DESARMA_TARIMA
70267,23155979,1,2B347696GN100,EC57 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,S05H042719600,NOA,TAR402408-EMBS2_CSIN,TAR402408-APT
69632,23155319,3,4A619437PD300,EC53 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,68055965SD,NOA,TAR478804-EMBSL1_CHU,TAR478804-APT
70592,23155319,3,4A619437PD300,EC53 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,68055965SD,NOA,TAR478805-EMBSL1_CHU,TAR478805-APT
69630,23155319,4,4A619437PD300,EC53 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,68055965SD,NOA,TAR478804-EMBSL1_CHU,TAR478804-APT
69629,23155261,2,4A613955CD300,EC68 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180691AA,NOA,TAR478191-APT,NaN
...,...,...,...,...,...,...,...,...
120,20810912,2,3B719986CD300,EC53 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180344AA,NOA,TAR418764-APT,TAR418764-APT
119,20810909,1,3B719986CD300,EC53 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,180344AA,NOA,TAR418764-APT,TAR418764-APT
592,20810758,3,4A519313CD300,EC60 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,2031850D,NOA,TAR418751-APT,TAR418751-APT
590,20810758,2,4A519313CD300,EC60 - CINTA / PAPEL / OJO HORIZONTAL / TERRESTRE,2031850D,NOA,TAR418751-APT,TAR418751-APT


Se ve que se arman tarimas con cintas de diferente material padre 

In [33]:
grupos = df_desarme.groupby('ARMA_TARIMA')['MATERIAL_PADRE'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada ARMA_TARIMA siempre tiene el mismo valor de MATERIAL_PADRE en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} ARMA_TARIMA(s) tienen más de un valor distinto en MATERIAL_PADRE:")
    print(inconsistentes)

NO SE CUMPLE: 179 ARMA_TARIMA(s) tienen más de un valor distinto en MATERIAL_PADRE:
ARMA_TARIMA
TAR343154-APT           2
TAR382411-EMBSL2_JUV    3
TAR387395-APT           3
TAR394142-EMBSL2_JUV    2
TAR400077-APT           2
                       ..
TAR478415-APT           2
TAR478422-APT           2
TAR478687-APT           2
TAR478950-APT           2
TAR480601-APT           2
Name: MATERIAL_PADRE, Length: 179, dtype: int64


Se ve que se arman tarimas con cintas de diferente producto

In [34]:
grupos = df_desarme.groupby('ARMA_TARIMA')['NUMERO_PARTE_HIJO'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada ARMA_TARIMA siempre tiene el mismo valor de NUMERO_PARTE_HIJO en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} ARMA_TARIMA(s) tienen más de un valor distinto en NUMERO_PARTE_HIJO:")
    print(inconsistentes)

NO SE CUMPLE: 277 ARMA_TARIMA(s) tienen más de un valor distinto en NUMERO_PARTE_HIJO:
ARMA_TARIMA
TAR387441-APT           2
TAR406342-APT           2
TAR411868-EMBSL2_CHU    2
TAR416272-APT           2
TAR416365-EMBSL2_CHU    2
                       ..
TAR478493-EMBS2_CSIN    2
TAR478515-EMBS2_CSIN    2
TAR479034-EMBSL2_CHU    2
TAR479309-APT           2
TAR480601-APT           2
Name: NUMERO_PARTE_HIJO, Length: 277, dtype: int64


Se crea un datafreame con unicamente las cintas que se armo tarima y no se desarmo la tarima

In [40]:
df_pq_desarme = df_desarme
variables_compaDesarme = [
                    'MATERIAL_PADRE',
                    'NUMERO_PARTE_HIJO',
                    'ARMA_TARIMA',
                    'DESARMA_TARIMA']



df_pq_desarme = df_pq_desarme[variables_compaDesarme]
df_pq_desarme = df_pq_desarme.sort_values('ARMA_TARIMA', ascending=False)
df_pq_desarme = df_pq_desarme[df_pq_desarme['DESARMA_TARIMA'].isna()]
df_pq_desarme


,MATERIAL_PADRE,NUMERO_PARTE_HIJO,ARMA_TARIMA,DESARMA_TARIMA
70584,4A620748CD300,180326AB,TAR480769-APT,NaN
70583,4A620748CD300,180326AB,TAR480769-APT,NaN
70580,4A620888CD300,180326AB,TAR480763-APT,NaN
70579,4A620888CD300,180326AB,TAR480763-APT,NaN
70523,4A610636CD300,MCI175AB.,TAR480390-APT,NaN
...,...,...,...,...
57490,4A427763CD300,2031795A,TAR382902-APT,NaN
57488,4A427763CD300,2031795A,TAR382901-APT,NaN
57489,4A427763CD300,2031795A,TAR382901-APT,NaN
34681,4A272719JP201,NaN,TAR343154-APT,NaN


Volvemos a ver si existen aun tarimas que se armaron (pero no se desarmaron despues) que tengan diferente material padre, vemos que aun hay

In [36]:
grupos = df_pq_desarme.groupby('ARMA_TARIMA')['MATERIAL_PADRE'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada ARMA_TARIMA siempre tiene el mismo valor de MATERIAL_PADRE en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} ARMA_TARIMA(s) tienen más de un valor distinto en MATERIAL_PADRE:")
    print(inconsistentes)

NO SE CUMPLE: 122 ARMA_TARIMA(s) tienen más de un valor distinto en MATERIAL_PADRE:
ARMA_TARIMA
TAR343154-APT           2
TAR394142-EMBSL2_JUV    2
TAR400077-APT           2
TAR411105-EMBSL1_CHU    2
TAR411299-EMBSL2_JUV    2
                       ..
TAR477970-EMBSL2_CHU    3
TAR478363-EMBSL1_CHU    2
TAR478415-APT           2
TAR478687-APT           2
TAR478950-APT           2
Name: MATERIAL_PADRE, Length: 122, dtype: int64


Ahora vemos si existen tarimas que se armaron y no se desarmaron despues que tengan cintas que vengan de diferente producto. vemos que hay muchisimas menos (de 277 tarimas que se armaron con productos distintos la mayoria se desarmó y solo quedan 3)

Esto prueba que en una misma tarima no pueden ir cintas que vengan de difernte producto

In [37]:
grupos = df_pq_desarme.groupby('ARMA_TARIMA')['NUMERO_PARTE_HIJO'].nunique()


inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada ARMA_TARIMA siempre tiene el mismo valor de NUMERO_PARTE_HIJO en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} ARMA_TARIMA(s) tienen más de un valor distinto en NUMERO_PARTE_HIJO:")
    print(inconsistentes)

NO SE CUMPLE: 3 ARMA_TARIMA(s) tienen más de un valor distinto en NUMERO_PARTE_HIJO:
ARMA_TARIMA
TAR469725-APT    2
TAR472319-APT    2
TAR473389-APT    2
Name: NUMERO_PARTE_HIJO, dtype: int64


In [41]:
df_dondeViene = df
df_dondeViene = df_dondeViene[df_dondeViene['DESARMA_TARIMA'].isna()]


In [ ]:
df_dondeViene

In [42]:
grupos = df_dondeViene.groupby('ARMA_TARIMA')['CONSIGNATARIO'].nunique()

# Grupos donde ID_ODE tiene más de un valor distinto de C_PEDIDO
inconsistentes = grupos[grupos > 1]

if inconsistentes.empty:
    print("CONFIRMADO: Cada ARMA_TARIMA siempre tiene el mismo valor de CONSIGNATARIO en todas sus filas.")
else:
    print(f"NO SE CUMPLE: {len(inconsistentes)} ARMA_TARIMA(s) tienen más de un valor distinto en CONSIGNATARIO:")
    print(inconsistentes)

CONFIRMADO: Cada ARMA_TARIMA siempre tiene el mismo valor de CONSIGNATARIO en todas sus filas.


In [43]:
# ============================================================
# BLOQUE 1: Distribución de tarimas válidas por NUMERO_PARTE_HIJO
# Queremos saber si hay concentración (pocos números de parte
# con muchas tarimas) o distribución uniforme
# ============================================================

# Filtramos: tarimas no desarmadas y mismo MATERIAL_PADRE dentro de cada tarima
tarimas_validas = df[df['DESARMA_TARIMA'].isna()].copy()

# Identificar tarimas que tienen más de un MATERIAL_PADRE (las excepciones a eliminar)
tarimas_multi_padre = (
    tarimas_validas.groupby('ARMA_TARIMA')['MATERIAL_PADRE']
    .nunique()
)
tarimas_a_eliminar = tarimas_multi_padre[tarimas_multi_padre > 1].index

# Quedarnos solo con tarimas "limpias"
tarimas_validas = tarimas_validas[~tarimas_validas['ARMA_TARIMA'].isin(tarimas_a_eliminar)]

print(f"Tarimas válidas (no desarmadas, mismo MATERIAL_PADRE): {tarimas_validas['ARMA_TARIMA'].nunique()}")
print(f"NUMERO_PARTE_HIJO distintos en tarimas válidas: {tarimas_validas['NUMERO_PARTE_HIJO'].nunique()}")

# Contar cuántas tarimas únicas tiene cada número de parte
tarimas_por_parte = (
    tarimas_validas.groupby('NUMERO_PARTE_HIJO')['ARMA_TARIMA']
    .nunique()
    .sort_values(ascending=False)
    .reset_index()
)
tarimas_por_parte.columns = ['NUMERO_PARTE_HIJO', 'n_tarimas']

# Distribución por rangos
bins = [0, 1, 3, 5, 10, 20, 50, float('inf')]
labels = ['1', '2-3', '4-5', '6-10', '11-20', '21-50', '50+']
tarimas_por_parte['rango'] = pd.cut(tarimas_por_parte['n_tarimas'], bins=bins, labels=labels)

resumen = tarimas_por_parte.groupby('rango', observed=True)['NUMERO_PARTE_HIJO'].count()
resumen_pct = (resumen / resumen.sum() * 100).round(1)

print("\nDistribución de números de parte por cantidad de tarimas válidas:")
print(pd.DataFrame({'n_numeros_parte': resumen, 'porcentaje': resumen_pct}))

Tarimas válidas (no desarmadas, mismo MATERIAL_PADRE): 7680
NUMERO_PARTE_HIJO distintos en tarimas válidas: 652

Distribución de números de parte por cantidad de tarimas válidas:
       n_numeros_parte  porcentaje
rango                             
1                  117        17.9
2-3                151        23.2
4-5                 87        13.3
6-10               134        20.6
11-20               81        12.4
21-50               59         9.0
50+                 23         3.5


In [44]:
# ============================================================
# BLOQUE 2: ¿Cuántos NUMERO_PARTE_HIJO ya tienen valores conocidos
# en PIEZAS_POR_PAQ_HIJO y PESO_MAX_DESPACHO_HIJO?
# Solo necesitamos estimar los que tienen NOA
# ============================================================

# Reemplazar 'NOA' por NaN para tratarlos como faltantes
df_check = df.copy()
df_check['PIEZAS_POR_PAQ_HIJO'] = df_check['PIEZAS_POR_PAQ_HIJO'].replace('NOA', np.nan)
df_check['PESO_MAX_DESPACHO_HIJO'] = df_check['PESO_MAX_DESPACHO_HIJO'].replace('NOA', np.nan)

# Para cada NUMERO_PARTE_HIJO, ver si tiene al menos un valor conocido
resumen_partes = df_check.groupby('NUMERO_PARTE_HIJO').agg(
    tiene_piezas=('PIEZAS_POR_PAQ_HIJO', lambda x: x.notna().any()),
    tiene_peso_max=('PESO_MAX_DESPACHO_HIJO', lambda x: x.notna().any()),
    total_cintas=('ARMA_TARIMA', 'count')
).reset_index()

print("=== PIEZAS_POR_PAQ_HIJO ===")
print(resumen_partes['tiene_piezas'].value_counts().rename({True: 'Tiene valor conocido', False: 'Solo NOA'}))

print("\n=== PESO_MAX_DESPACHO_HIJO ===")
print(resumen_partes['tiene_peso_max'].value_counts().rename({True: 'Tiene valor conocido', False: 'Solo NOA'}))

print("\n=== Combinación de ambas ===")
resumen_partes['categoria'] = 'Solo NOA en ambas'
resumen_partes.loc[resumen_partes['tiene_piezas'] & resumen_partes['tiene_peso_max'], 'categoria'] = 'Ambas conocidas'
resumen_partes.loc[resumen_partes['tiene_piezas'] & ~resumen_partes['tiene_peso_max'], 'categoria'] = 'Solo piezas conocida'
resumen_partes.loc[~resumen_partes['tiene_piezas'] & resumen_partes['tiene_peso_max'], 'categoria'] = 'Solo peso conocido'

print(resumen_partes['categoria'].value_counts())

=== PIEZAS_POR_PAQ_HIJO ===
tiene_piezas
Solo NOA                650
Tiene valor conocido     86
Name: count, dtype: int64

=== PESO_MAX_DESPACHO_HIJO ===
tiene_peso_max
Solo NOA                587
Tiene valor conocido    149
Name: count, dtype: int64

=== Combinación de ambas ===
categoria
Solo NOA en ambas       544
Solo peso conocido      106
Ambas conocidas          43
Solo piezas conocida     43
Name: count, dtype: int64


In [45]:
# ============================================================
# BLOQUE 3: Para los que SÍ tienen valor conocido,
# ¿es consistente ese valor o varía dentro del mismo número de parte?
# Esto nos dice qué tan confiables son los valores existentes
# ============================================================

# Solo filas con valor conocido en PIEZAS_POR_PAQ_HIJO
piezas_conocidas = df_check[df_check['PIEZAS_POR_PAQ_HIJO'].notna()]

consistencia_piezas = piezas_conocidas.groupby('NUMERO_PARTE_HIJO')['PIEZAS_POR_PAQ_HIJO'].nunique()
print("=== Consistencia de PIEZAS_POR_PAQ_HIJO ===")
print("Valores distintos por número de parte:")
print(consistencia_piezas.value_counts().rename_axis('n_valores_distintos').reset_index(name='n_numeros_parte'))

# Solo filas con valor conocido en PESO_MAX_DESPACHO_HIJO
peso_conocido = df_check[df_check['PESO_MAX_DESPACHO_HIJO'].notna()]

consistencia_peso = peso_conocido.groupby('NUMERO_PARTE_HIJO')['PESO_MAX_DESPACHO_HIJO'].nunique()
print("\n=== Consistencia de PESO_MAX_DESPACHO_HIJO ===")
print("Valores distintos por número de parte:")
print(consistencia_peso.value_counts().rename_axis('n_valores_distintos').reset_index(name='n_numeros_parte'))

=== Consistencia de PIEZAS_POR_PAQ_HIJO ===
Valores distintos por número de parte:
   n_valores_distintos  n_numeros_parte
0                    1               85
1                    2                1

=== Consistencia de PESO_MAX_DESPACHO_HIJO ===
Valores distintos por número de parte:
   n_valores_distintos  n_numeros_parte
0                    1              145
1                    2                4


In [11]:
# ============================================================
# BLOQUE 4: Homogeneidad interna de NUMERO_PARTE_HIJO
# Queremos saber si las cintas del mismo número de parte
# tienen dimensiones consistentes o muy variables
# ============================================================

# Asegurarnos de que df_check existe (puede venir del BLOQUE 2 o se recrea aquí)
if 'df_check' not in dir():
    df_check = df.copy()
    df_check['PIEZAS_POR_PAQ_HIJO'] = df_check['PIEZAS_POR_PAQ_HIJO'].replace('NOA', np.nan)
    df_check['PESO_MAX_DESPACHO_HIJO'] = df_check['PESO_MAX_DESPACHO_HIJO'].replace('NOA', np.nan)

# Asegurarnos de trabajar con valores numéricos
df_check['ANCHO_HIJO'] = pd.to_numeric(df_check['ANCHO_HIJO'], errors='coerce')
df_check['ESPESOR_HIJO'] = pd.to_numeric(df_check['ESPESOR_HIJO'], errors='coerce')

# Para cada número de parte, calcular rango (max - min) de cada dimensión
homogeneidad = df_check.groupby('NUMERO_PARTE_HIJO').agg(
    n_cintas=('ANCHO_HIJO', 'count'),
    ancho_min=('ANCHO_HIJO', 'min'),
    ancho_max=('ANCHO_HIJO', 'max'),
    ancho_rango=('ANCHO_HIJO', lambda x: x.max() - x.min()),
    ancho_valores_distintos=('ANCHO_HIJO', 'nunique'),
    espesor_min=('ESPESOR_HIJO', 'min'),
    espesor_max=('ESPESOR_HIJO', 'max'),
    espesor_rango=('ESPESOR_HIJO', lambda x: x.max() - x.min()),
    espesor_valores_distintos=('ESPESOR_HIJO', 'nunique'),
).reset_index()

# ¿Cuántos números de parte tienen exactamente 1 valor de ancho y 1 de espesor?
perfectamente_homogeneos = homogeneidad[
    (homogeneidad['ancho_valores_distintos'] == 1) & 
    (homogeneidad['espesor_valores_distintos'] == 1)
]
print(f"Números de parte con ancho Y espesor completamente fijos: "
      f"{len(perfectamente_homogeneos)} de {len(homogeneidad)} "
      f"({len(perfectamente_homogeneos)/len(homogeneidad)*100:.1f}%)")

# Distribución del rango de ancho
print("\n=== Variación de ANCHO_HIJO por número de parte (mm) ===")
bins_ancho = [-1, 0, 1, 5, 10, float('inf')]
labels_ancho = ['Fijo (0)', 'Hasta 1mm', '1-5mm', '5-10mm', 'Más de 10mm']
homogeneidad['rango_ancho_cat'] = pd.cut(homogeneidad['ancho_rango'], bins=bins_ancho, labels=labels_ancho)
print(homogeneidad['rango_ancho_cat'].value_counts().sort_index())

# Distribución del rango de espesor
print("\n=== Variación de ESPESOR_HIJO por número de parte (mm) ===")
bins_esp = [-1, 0, 0.1, 0.5, 1, float('inf')]
labels_esp = ['Fijo (0)', 'Hasta 0.1mm', '0.1-0.5mm', '0.5-1mm', 'Más de 1mm']
homogeneidad['rango_espesor_cat'] = pd.cut(homogeneidad['espesor_rango'], bins=bins_esp, labels=labels_esp)
print(homogeneidad['rango_espesor_cat'].value_counts().sort_index())

# Ver los números de parte más heterogéneos (los problemáticos)
print("\n=== Top 10 números de parte con mayor variación en ANCHO_HIJO ===")
print(homogeneidad.nlargest(10, 'ancho_rango')[
    ['NUMERO_PARTE_HIJO', 'n_cintas', 'ancho_min', 'ancho_max', 'ancho_rango', 'espesor_rango']
].to_string(index=False))

Números de parte con ancho Y espesor completamente fijos: 176 de 736 (23.9%)

=== Variación de ANCHO_HIJO por número de parte (mm) ===
rango_ancho_cat
Fijo (0)       209
Hasta 1mm      461
1-5mm           25
5-10mm           7
Más de 10mm     34
Name: count, dtype: int64

=== Variación de ESPESOR_HIJO por número de parte (mm) ===
rango_espesor_cat
Fijo (0)       223
Hasta 0.1mm    494
0.1-0.5mm       19
0.5-1mm          0
Más de 1mm       0
Name: count, dtype: int64

=== Top 10 números de parte con mayor variación en ANCHO_HIJO ===
            NUMERO_PARTE_HIJO  n_cintas  ancho_min  ancho_max  ancho_rango  espesor_rango
 Z CINTA 3' C.22 G33 G90 ANCH       465      60.11     345.15       285.04         0.0630
                     2034298F        12     300.00     570.50       270.50         0.0300
                     2032499I       129     169.80     410.10       240.30         0.0400
                     180636AA         4     200.00     370.86       170.86         0.0100
            

In [12]:
# ============================================================
# BLOQUE 5: ¿Qué tan relevantes son los números de parte
# con ancho variable (>1mm)? ¿Cuántas cintas representan?
# ============================================================

# Identificar números de parte problemáticos (variación de ancho > 1mm)
partes_problematicas = homogeneidad[homogeneidad['ancho_rango'] > 1]['NUMERO_PARTE_HIJO']
partes_ok = homogeneidad[homogeneidad['ancho_rango'] <= 1]['NUMERO_PARTE_HIJO']

total_cintas = len(df_check)
cintas_problematicas = df_check[df_check['NUMERO_PARTE_HIJO'].isin(partes_problematicas)]
cintas_ok = df_check[df_check['NUMERO_PARTE_HIJO'].isin(partes_ok)]

print("=== Impacto de números de parte con ancho variable (>1mm) ===")
print(f"Números de parte problemáticos: {len(partes_problematicas)} de {len(homogeneidad)} ({len(partes_problematicas)/len(homogeneidad)*100:.1f}%)")
print(f"Cintas en partes problemáticas: {len(cintas_problematicas):,} de {total_cintas:,} ({len(cintas_problematicas)/total_cintas*100:.1f}%)")
print(f"Cintas en partes OK:            {len(cintas_ok):,} de {total_cintas:,} ({len(cintas_ok)/total_cintas*100:.1f}%)")

# De las cintas problemáticas, ¿cuántas están en tarimas desarmadas vs no desarmadas?
print("\n=== Dentro de las cintas problemáticas ===")
print(cintas_problematicas['DESARMA_TARIMA'].isna().value_counts()
      .rename({True: 'No desarmadas', False: 'Desarmadas'}))

# ¿Las partes problemáticas tienen más desarmes que las OK?
tasa_desarme_problematicas = cintas_problematicas['DESARMA_TARIMA'].notna().mean()
tasa_desarme_ok = cintas_ok['DESARMA_TARIMA'].notna().mean()

print(f"\nTasa de desarme en partes problemáticas: {tasa_desarme_problematicas*100:.1f}%")
print(f"Tasa de desarme en partes OK:            {tasa_desarme_ok*100:.1f}%")

=== Impacto de números de parte con ancho variable (>1mm) ===
Números de parte problemáticos: 66 de 736 (9.0%)
Cintas en partes problemáticas: 4,110 de 27,743 (14.8%)
Cintas en partes OK:            23,449 de 27,743 (84.5%)

=== Dentro de las cintas problemáticas ===
DESARMA_TARIMA
No desarmadas    3167
Desarmadas        943
Name: count, dtype: int64

Tasa de desarme en partes problemáticas: 22.9%
Tasa de desarme en partes OK:            19.1%


se va a excluir las cintas que vienen del mismo numero parte hijo y tienen dimensiones muy diferentes